### PR TITLE
Query: Tuples are not translatable to Sql. Client Eval them

### LINKED ISSUE #9424: .Contains on a tuple collection doesn't work
This works in 1.1.2.

Add this test to SimpleQueryTestBase.ResultOperators.cs:

```C#
[ConditionalFact]
public virtual void Contains_with_local_tuple_array_closure()
{
    var ids = new[] { Tuple.Create(1, 2), Tuple.Create(10248, 11) };

    AssertQuery<OrderDetail>(
        cd => cd.Where(c => ids.Contains(new Tuple<int, int>(c.OrderID, c.ProductID))), entryCount: 1);

    ids = new[] { Tuple.Create(1, 2) };

    AssertQuery<OrderDetail>(
        cd => cd.Where(c => ids.Contains(new Tuple<int, int>(c.OrderID, c.ProductID))));
}
```

>System.InvalidOperationException : No mapping to a relational type can be found for the CLR type 'Expression[]'.
	at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
	at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
	at System.Threading.Tasks.Task.Wait()
	C:\enlistments\Universe\EntityFramework\src\EFCore.Specification.Tests\Query\QueryTestBase.cs(83,0): at Microsoft.EntityFrameworkCore.Query.QueryTestBase`1.AssertQuery[TItem1](Func`2 query, Func`2 elementSorter, Action`2 elementAsserter, Boolean assertOrder, Int32 entryCount)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Specification.Tests\Query\SimpleQueryTestBase.ResultOperators.cs(678,0): at Microsoft.EntityFrameworkCore.Query.SimpleQueryTestBase`1.Contains_with_local_tuple_array_closure()
	C:\enlistments\Universe\EntityFramework\test\EFCore.SqlServer.FunctionalTests\Query\SimpleQuerySqlServerTest.ResultOperators.cs(13,0): at Microsoft.EntityFrameworkCore.Query.SimpleQuerySqlServerTest.Contains_with_local_tuple_array_closure()
	----- Inner Stack Trace -----
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Storage\RelationalTypeMapperExtensions.cs(80,0): at Microsoft.EntityFrameworkCore.Storage.RelationalTypeMapperExtensions.GetMapping(IRelationalTypeMapper typeMapper, Type clrType)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Storage\RelationalTypeMapperExtensions.cs(27,0): at Microsoft.EntityFrameworkCore.Storage.RelationalTypeMapperExtensions.GetMappingForValue(IRelationalTypeMapper typeMapper, Object value)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Sql\DefaultQuerySqlGenerator.cs(589,0): at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.GetTypeMapping(Object value)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Sql\DefaultQuerySqlGenerator.cs(1448,0): at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.VisitConstant(ConstantExpression constantExpression)
	at System.Linq.Expressions.ConstantExpression.Accept(ExpressionVisitor visitor)
	at Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Sql\DefaultQuerySqlGenerator.cs(725,0): at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.VisitIn(InExpression inExpression)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Expressions\InExpression.cs(106,0): at Microsoft.EntityFrameworkCore.Query.Expressions.InExpression.Accept(ExpressionVisitor visitor)
	at Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Sql\DefaultQuerySqlGenerator.cs(359,0): at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.GeneratePredicate(Expression predicate)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Sql\DefaultQuerySqlGenerator.cs(236,0): at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.VisitSelect(SelectExpression selectExpression)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Expressions\SelectExpression.cs(1041,0): at Microsoft.EntityFrameworkCore.Query.Expressions.SelectExpression.Accept(ExpressionVisitor visitor)
	at Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Sql\DefaultQuerySqlGenerator.cs(130,0): at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.GenerateSql(IReadOnlyDictionary`2 parameterValues)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Internal\ShaperCommandContext.cs(118,0): at Microsoft.EntityFrameworkCore.Query.Internal.ShaperCommandContext.GetRelationalCommand(IReadOnlyDictionary`2 parameters)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Internal\QueryingEnumerable.cs(131,0): at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(Boolean buffer)
	C:\enlistments\Universe\EntityFramework\src\EFCore\Extensions\ExecutionStrategyExtensions.cs(337,0): at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.<>c__DisplayClass12_0`2.<Execute>b__0(DbContext c, TState s)
	C:\enlistments\Universe\EntityFramework\src\EFCore.SqlServer\Storage\Internal\SqlServerExecutionStrategy.cs(39,0): at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
	C:\enlistments\Universe\EntityFramework\src\EFCore\Extensions\ExecutionStrategyExtensions.cs(335,0): at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.Execute[TState,TResult](IExecutionStrategy strategy, Func`2 operation, Func`2 verifySucceeded, TState state)
	C:\enlistments\Universe\EntityFramework\src\EFCore\Extensions\ExecutionStrategyExtensions.cs(285,0): at Microsoft.EntityFrameworkCore.ExecutionStrategyExtensions.Execute[TState,TResult](IExecutionStrategy strategy, TState state, Func`2 operation)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Relational\Query\Internal\QueryingEnumerable.cs(75,0): at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
	C:\enlistments\Universe\EntityFramework\src\EFCore\Query\Internal\LinqOperatorProvider.cs(173,0): at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.<_TrackEntities>d__17`2.MoveNext()
	C:\enlistments\Universe\EntityFramework\src\EFCore\Query\Internal\LinqOperatorProvider.cs(137,0): at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
	at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
	at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
	C:\enlistments\Universe\EntityFramework\src\EFCore.Specification.Tests\TestUtilities\QueryAsserter.cs(333,0): at Microsoft.EntityFrameworkCore.TestUtilities.QueryAsserter`1.<AssertQuery>d__18`1.MoveNext()

### PR DESCRIPTION
**Pull request by smitpatel**
Fixes #9424

### CODE REVIEW

File: src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs [hunk 1]
```diff
@@ -866,6 +866,20 @@ public virtual void Contains_top_level()
         }
 
         [ConditionalFact]
+        public virtual void Contains_with_local_tuple_array_closure()
+        {
+            var ids = new[] { Tuple.Create(1, 2), Tuple.Create(10248, 11) };
+
+            AssertQuery<OrderDetail>(
+                cd => cd.Where(c => ids.Contains(new Tuple<int, int>(c.OrderID, c.ProductID))), entryCount: 1);
```
Comment (anpete) [inline src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs]: cd -> od
c -> od

What happens with normal anonymous type for this test?

File: src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -845,7 +845,14 @@ protected override Expression VisitNew(NewExpression expression)
...
 
-            if (expression.Members != null
+            /* Tuples are converted to NewExpression with Tuple type in expression tree
+             * We cannot translate tuples to Sql
+             * Since there are 8 different generic Tuple types doing string comparison to exclude them
+             * If this breaks then we can move to pure reflection based comparison.
+             */
+            if ((!expression.Type.IsGenericType
+                 || !string.Equals(expression.Type.GetGenericTypeDefinition().DisplayName(), typeof(Tuple).DisplayName()))
+                && expression.Members != null
```
Comment (anpete) [inline src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs]: Can we do the anonymous type expansion?

File: src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs (hunk 1 above)
Comment (smitpatel) [inline src/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs]: It also fails. Needs different solution.

**Comment by smitpatel** [comment]
Updated

File: src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs [hunk 3]
```diff
@@ -907,7 +907,7 @@ protected override Expression VisitSubQuery(SubQueryExpression expression)
                     || fromExpression.NodeType == ExpressionType.NewArrayInit)
                 {
                     var containsItem = Visit(contains.Item)?.RemoveConvert();
-                    if (containsItem != null)
+                    if (containsItem != null && !containsItem.Type.Equals(typeof(Expression[])))
```
Comment (anpete) [inline src/EFCore.Relational/Query/ExpressionVisitors/SqlTranslatingExpressionVisitor.cs]: == :trollface:

**Comment by smitpatel** [comment]
Added more tests. We are not translating equality in Where for anonymous/tuple.

**Comment by smitpatel** [comment]
@anpete - Ping!

**Pull request merged**